packages/protobuf/test/scenarios.spec.ts

```diff
@@ -0,0 +1,217 @@
+
+          const diagnostics = processedDiagnostics.join("\n");
+
+          assert.strictEqual(diagnostics, expectedDiagnostics, "expected equivalent diagnostics");
+
+          if (err) throw err;
+        }
+      });
+  }
+});
+
+interface EmitResult {
+  files: Record<string, string>;
+  diagnostics: string[];
+}
+
+async function doEmit(files: Record<string, string>): Promise<EmitResult> {
+  const baseOutputPath = resolveVirtualPath("test-output/");
+
+  const host = await createTestHost({
+    libraries: [TypeSpecProtobufTestLibrary],
+  });
+
+  for (const [fileName, content] of Object.entries(files)) {
+    host.addTypeSpecFile(fileName, content);
+  }
+
+  const [, diagnostics] = await host.compileAndDiagnose("main.tsp", {
+    outputDir: baseOutputPath,
+    noEmit: false,
+    emitters: {
+      "@typespec/protobuf": {},
+    },
+  });
+
+  return {
+    files: Object.fromEntries(
+      [...host.fs.entries()]
+        .filter(([name]) => name.startsWith(baseOutputPath))
+        .map(([name, value]) => [name.replace(baseOutputPath, ""), value])
+    ),
+    diagnostics: diagnostics.map(formatDiagnostic),
+  };
+}
+
+function assertFilesAsExpected(
+  outputFiles: Record<string, string>,
+  expectedFiles: Record<string, string>
+) {
+  for (const fn of Object.keys(expectedFiles)) {
+    assert.ok(
+      Object.prototype.hasOwnProperty.call(outputFiles, fn),
+      `expected file ${fn} was not produced`
+    );
+  }
+
+  for (const [fn, content] of Object.entries(outputFiles)) {
+    const expectedContent = expectedFiles[fn];
+
+    assert.ok(expectedContent, `output file ${fn} has no corresponding expectation`);
+
+    assert.strictEqual(content, expectedContent);
+  }
+}
+
+async function writeExpectationDirectory(
+  expectationDirectory: string,
+  outputFiles: Record<string, string>
+) {
+  const fileEntries = Object.entries(outputFiles);
+
+  // It'll be annoying to fiddle with .gitkeep files, so let's omit the `output` directory if it's empty.
+  if (fileEntries.length === 0) {
+    return;
+  }
+
+  await fs.promises.rm(expectationDirectory, { recursive: true, force: true });
+
+  await fs.promises.mkdir(expectationDirectory);
+
+  for (const [fn, content] of fileEntries) {
+    const fullPath = path.join(expectationDirectory, fn);
+    await fs.promises.mkdir(path.dirname(fullPath), { recursive: true });
+    await fs.promises.writeFile(fullPath, content);
+  }
+}
+
+async function readdirRecursive(dir: string, base: string = dir): Promise<Record<string, string>> {
+  const res: Record<string, string> = {};
+
+  for (const entry of (await fs.promises.readdir(dir)).map((e) => path.join(dir, e))) {
+    const stat = await fs.promises.stat(entry);
+
+    if (stat.isDirectory()) {
+      for (const [name, content] of Object.entries(await readdirRecursive(entry, base))) {
+        res[name] = content;
+      }
+    } else if (stat.isFile()) {
+      const content = (await fs.promises.readFile(entry)).toString("utf-8");
```

Comment:
```suggestion
      const content = (await readFile(entry)).toString("utf-8");
```
would`import from "fs/promises"` instead of importing all of `fs`